src/sema/using.rs

```diff
@@ -113,19 +116,145 @@ pub(crate) fn using_decl(
+                                        loc,
+                                        format!("definition of '{function_name}'"),
+                                    ));
+                                    continue;
+                                }
+                            }
+                        };
+
+                        if func.params.len() != oper.args()
+                            || func.params.iter().any(|param| param.ty != *ty)
+                        {
                             diagnostics.push(Diagnostic::error_with_note(
-                                function_name.loc,
-                                format!("function cannot be used since first argument is '{}' rather than the required '{}'", func.params[0].ty.to_string(ns), ty.to_string(ns)),
+                                using_function.loc,
+                                format!(
+                                    "user defined operator function for '{}' must have {} arguments of type {}",
+                                    oper, oper.args(), ty.to_string(ns)
+                                ),
+                                loc,
+                                format!("definition of '{function_name}'"),
+                            ));
+                            continue;
+                        }
+
+                        if oper.is_comparison() {
+                            if func.returns.len() != 1 || func.returns[0].ty != Type::Bool {
+                                diagnostics.push(Diagnostic::error_with_note(
+                                    using_function.loc,
+                                    format!(
+                                        "user defined operator function for '{}' must have one bool return type",
+                                        oper,
+                                    ),
+                                    loc,
+                                    format!("definition of '{function_name}'"),
+                                ));
+                                continue;
+                            }
+                        } else if func.returns.len() != 1 || func.returns[0].ty != *ty {
+                            diagnostics.push(Diagnostic::error_with_note(
+                                using_function.loc,
+                                    format!(
+                                        "user defined operator function for '{}' must have single return type {}",
+                                        oper, ty.to_string(ns)
+                                    ),
+                                    loc,
+                                    format!("definition of '{function_name}'"),
+                                ));
+                            continue;
+                        }
+
+                        if !matches!(func.mutability, Mutability::Pure(_)) {
```

Comment:
All this if-conditions can be an else-if chain, although this is not strictly necessary.

Reply:
It is a good idea, but when I tried it, the readability was not improved. It makes the flow of the code inconsistent with the preceding code.